From 7b73ca7d48466953c6ce79c2e9392f518c21ae49 Mon Sep 17 00:00:00 2001
From: rikardo <rikard.olander@hostmobility.com>
Date: Fri, 24 Jan 2025 08:28:53 +0000
Subject: [PATCH 3/4] HMM customization

v2. Fix m4boot and move eeprom code block to better location

---
 board/toradex/common/tdx-common.c | 99 ++++++++++++++++++++++++-------
 include/configs/verdin-am62.h     | 41 +++++++------
 2 files changed, 99 insertions(+), 41 deletions(-)

diff --git a/board/toradex/common/tdx-common.c b/board/toradex/common/tdx-common.c
index d144914397..44ef87d70a 100644
--- a/board/toradex/common/tdx-common.c
+++ b/board/toradex/common/tdx-common.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0+
 /*
  * Copyright (c) 2016 Toradex, Inc.
+ * Copyright 2024 SETEK Systems AB
  */
 
 #include <common.h>
@@ -29,9 +30,7 @@ static char tdx_serial_str[SERIAL_STR_LEN + 1];
 static char tdx_board_rev_str[MODULE_VER_STR_LEN + MODULE_REV_STR_LEN + 1];
 
 #ifdef CONFIG_TDX_CFG_BLOCK_EXTRA
-static char tdx_car_serial_str[SERIAL_STR_LEN + 1];
-static char tdx_car_rev_str[MODULE_VER_STR_LEN + MODULE_REV_STR_LEN + 1];
-static const char *tdx_carrier_board_name;
+#include "../common/hmm_eeprom.h"
 #endif
 
 #if defined(CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG)
@@ -99,6 +98,8 @@ static const char *get_board_assembly(u16 ver_assembly)
 int show_board_info(void)
 {
 	unsigned char ethaddr[6];
+	uchar eeprom_data [0xffff];
+	int err = 1;
 
 	if (read_tdx_cfg_block()) {
 		printf("MISSING TORADEX CONFIG BLOCK\n");
@@ -121,27 +122,18 @@ int show_board_info(void)
 		       tdx_board_rev_str);
 		printf("Serial#: %s\n", tdx_serial_str);
 #ifdef CONFIG_TDX_CFG_BLOCK_EXTRA
-		if (read_tdx_cfg_block_carrier()) {
-			printf("MISSING TORADEX CARRIER CONFIG BLOCKS\n");
-			try_migrate_tdx_cfg_block_carrier();
-		} else {
-			tdx_carrier_board_name =
-				get_toradex_carrier_boards(tdx_car_hw_tag.prodid);
-
-			snprintf(tdx_car_serial_str, sizeof(tdx_car_serial_str),
-				 "%08u", tdx_car_serial);
-			snprintf(tdx_car_rev_str, sizeof(tdx_car_rev_str),
-				 "V%1d.%1d%s",
-				 tdx_car_hw_tag.ver_major,
-				 tdx_car_hw_tag.ver_minor,
-				 get_board_assembly(tdx_car_hw_tag.ver_assembly));
-
-			env_set("carrier_serial#", tdx_car_serial_str);
-			printf("Carrier: Toradex %s %s, Serial# %s\n",
-			       tdx_carrier_board_name,
-			       tdx_car_rev_str,
-			       tdx_car_serial_str);
+	printf("Running Host Monitor Mini board from SETEK Systems AB\n");
+	/* HMM get env from eeprom serial number, partnumber and mac address(eth2addr, eth3addr) and more */
+	err = get_environment_data_from_eeprom(eeprom_data);
+	if (err != 0) {
+		printf("Failed to get environment data from eeprom, error: %x\n", err);
+	} else {
+		err = set_environment_data_from_eeprom((char *)eeprom_data);
+		if (err != 0) {
+			printf("Failed to set environment data to environment area, error: %x\n", err);
 		}
+	}
+
 #endif
 	}
 
@@ -181,6 +173,17 @@ int g_dnl_bind_fixup(struct usb_device_descriptor *dev, const char *name)
 #if defined(CONFIG_OF_LIBFDT)
 int ft_common_board_setup(void *blob, struct bd_info *bd)
 {
+	/* Set serial number and part number from environment to device tree blob chosen.
+		serial-number = "23078888";
+		part-number = "hmp00xx"; 
+		product-number = hmxxx;.
+	*/
+	char *serial;
+	char *part;
+	char *product;
+	int offset;
+	int err = 1;
+	const char *path;
 	if (tdx_serial) {
 		fdt_setprop(blob, 0, "serial-number", tdx_serial_str,
 			    strlen(tdx_serial_str) + 1);
@@ -196,6 +199,56 @@ int ft_common_board_setup(void *blob, struct bd_info *bd)
 			    strlen(tdx_board_rev_str) + 1);
 	}
 
+	path = "/chosen";
+	offset = fdt_path_offset(blob, path);
+	if (offset < 0) {
+		printf("WARNING: Node %s not found in device tree.\n", path);
+		return err;
+	}
+
+	debug("ft_board_setup: config FDT\n");
+
+	serial = env_get("serial");
+	if (serial) {
+		err = fdt_setprop(blob, offset, "serial-number", serial,
+				  strlen(serial) + 1);
+
+		if (err < 0) {
+			printf("WARNING: could not set serial-number %s.\n",
+			       fdt_strerror(err));
+			return err;
+		}
+	} else {
+		printf("WARNING: could not read serial-number from env list.\n");
+	}
+	part = env_get("part_nr");
+	if (part) {
+		err = fdt_setprop(blob, offset, "part-number", part,
+				  strlen(part) + 1);
+
+		if (err < 0) {
+			printf("WARNING: could not set part-number %s.\n",
+			       fdt_strerror(err));
+			return err;
+		}
+	} else {
+		printf("WARNING: could not read part-number from env list.\n");
+	}
+	product = env_get("product_nr");
+	if (product) {
+		err = fdt_setprop(blob, offset, "product-number", product,
+				  strlen(product) + 1);
+
+		if (err < 0) {
+			printf("WARNING: could not set product-number %s.\n",
+			       fdt_strerror(err));
+			return err;
+		}
+	} else {
+		printf("WARNING: could not read product-number from env list.\n");
+	}
+
+
 	return 0;
 }
 #endif
diff --git a/include/configs/verdin-am62.h b/include/configs/verdin-am62.h
index a2cbcce33d..bf590c659a 100644
--- a/include/configs/verdin-am62.h
+++ b/include/configs/verdin-am62.h
@@ -3,6 +3,7 @@
  * Configuration header file for Verdin AM62 SoM
  *
  * Copyright 2023 Toradex - https://www.toradex.com/
+ * Copyright 2024 SETEK Systems AB
  */
 
 #ifndef __VERDIN_AM62_H
@@ -27,6 +28,7 @@
 #if CONFIG_TARGET_VERDIN_AM62_A53
 /* Enable Distro Boot */
 #define BOOT_TARGET_DEVICES(func) \
+	func(USB, usb, 0) \
 	func(MMC, mmc, 1) \
 	func(MMC, mmc, 0) \
 	func(DHCP, dhcp, na)
@@ -45,30 +47,33 @@
 	"ramdisk_addr_r ram " __stringify(RAMDISK_ADDR_R) " 0x8000000\0"
 
 /* Incorporate settings into the U-Boot environment */
+/*m4boot will run in variant of Toradex boot.scr file we just declare it here instead of in the boot.scr file */
 #define CFG_EXTRA_ENV_SETTINGS \
 	BOOTENV \
 	EXTRA_ENV_DFUARGS \
 	MEM_LAYOUT_ENV_SETTINGS \
 	"boot_scripts=boot.scr\0" \
+	"bsp_script=hmm_boot.scr\0" \
 	"boot_script_dhcp=boot.scr\0" \
 	"console=ttyS2\0" \
-	"fdt_board=dev\0" \
+	"fdt_board=hmm\0" \
+	"bootcmd_usb=run try_usb_flash;\0" \
+	"try_usb_flash=sleep 0.1; usb reset; if load usb 0 ${scriptaddr} ${bsp_script} ; then source ${scriptaddr}; else echo UPDATE ERROR: no ${bsp_script} on first USB device. ;fi\0" \
+	"m4addr=0x82000000\0" \
+	"m4_firmware=lib/firmware/am62-mcu-m4f0_0-fw\0" \
+	"m4boot=if ext4size mmc ${root_devnum}:${root_part} ${m4_firmware}; then " \
+		"load mmc ${root_devnum}:${root_part} ${m4addr} ${m4_firmware}; " \
+			"if rproc init && rproc load 0 ${m4addr} ${filesize} && rproc start 0; then " \
+				"echo INFO: M4 firmware ${m4_firmware} successfully started; " \
+			"else " \
+				"echo ERROR: Failed to initialize or start M4 firmware; " \
+				"env set script_warning 'M4 firmware ${m4_firmware} initialization/start failed'; " \
+			"fi; " \
+		"else " \
+			"echo WARNING: File ${m4_firmware} not found on mmc ${root_devnum}:${root_part}, continue booting; " \
+			"env set script_warning 'File ${m4_firmware} not found on mmc ${root_devnum}:${root_part}, continue booting'; " \
+		"fi;\0" \
 	"setup=setenv setupargs console=tty1 console=${console},${baudrate} " \
-		"consoleblank=0 earlycon=ns16550a,mmio32,0x02800000\0" \
-	"update_tiboot3=askenv confirm Did you load tiboot3.bin (y/N)?; " \
-		"if test \"$confirm\" = \"y\"; then " \
-		"setexpr blkcnt ${filesize} + 0x1ff && setexpr blkcnt " \
-		"${blkcnt} / 0x200; mmc dev 0 1; mmc write ${loadaddr} 0x0 " \
-		"${blkcnt}; fi\0" \
-	"update_tispl=askenv confirm Did you load tispl.bin (y/N)?; " \
-		"if test \"$confirm\" = \"y\"; then " \
-		"setexpr blkcnt ${filesize} + 0x1ff && setexpr blkcnt " \
-		"${blkcnt} / 0x200; mmc dev 0 1; mmc write ${loadaddr} 0x400 " \
-		"${blkcnt}; fi\0" \
-	"update_uboot=askenv confirm Did you load u-boot.img (y/N)?; " \
-		"if test \"$confirm\" = \"y\"; then " \
-		"setexpr blkcnt ${filesize} + 0x1ff && setexpr blkcnt " \
-		"${blkcnt} / 0x200; mmc dev 0 1; mmc write ${loadaddr} 0x1400 " \
-		"${blkcnt}; fi\0"
+		"consoleblank=0 earlycon=ns16550a,mmio32,0x02800000\0"
 
-#endif /* __VERDIN_AM62_H */
+#endif /* __VERDIN_AM62_H */
\ No newline at end of file
-- 
2.34.1

